include/deal.II/base/mpi_consensus_algorithms.templates.h

```diff
@@ -209,14 +195,118 @@ namespace Utilities
       NBX<T1, T2>::all_locally_originated_receives_are_completed()
       {
 #ifdef DEAL_II_WITH_MPI
-        int        all_receive_requests_are_done;
-        const auto ierr = MPI_Testall(recv_requests.size(),
-                                      recv_requests.data(),
-                                      &all_receive_requests_are_done,
-                                      MPI_STATUSES_IGNORE);
-        AssertThrowMPI(ierr);
+        // We know that all requests have come in when we have pending
+        // messages from all targets with the right tag. We can check
+        // for pending messages with MPI_IProbe, which returns
+        // immediately with a return code that indicates whether
+        // it has found a message from a given process with a given
+        // tag
+        std::set<unsigned int> received_answers;
+        for (const unsigned int target : outstanding_answers)
```

Comment:
I am wondering if the code below is not equivalent to (if `MPI_Testany` is really the non-blocking equivalent of `MPI_Waitany` - see also  #https://mpi.deino.net/mpi_functions/MPI_Testany.html):
```cpp
if(counter == 0)
  return true;

MPI_Testany(targets.size(), requests, &index, &flag, &status);

if(flag == 0)
  return false;

MPI_Get_count(&status[index], MPI_BYTE, &message_size);

std::vector<T2> recv_buffer(message_size / sizeof(T2));

MPI_Recv(recv_buffer.data(), recv_buffer.size() * sizeof(T2), MPI_BYTE,
         targets[index], tag_deliver, this->comm, MPI_STATUS_IGNORE);

this->process.read_answer(target, recv_buffer);

counter--;

return counter !=0;
```
with `counter` initialized by `targets.size()`.

Reply:
The problem is that `MPI_Testany` requires an array of requests -- which one generally gets by calling something like `MPI_Irecv`, but for that I needed to know the buffer size :-( 

I think what *could* work is to call `MPI_Iprobe` where as the first argument I don't specify `target` (one of the outstanding answers), but `MPI_ANY_SOURCE`. This could work because I now deal with the resulting message right away, rather than deferring until we have received all of them. (If we defer, then `MPI_ANY_SOURCE` would not work: We would simply get a positive answer from `MPI_Iprobe` *every time* since we never pick up the message that has been delivered.

Would you like me to try this?